### ISSUE TITLE
Creating a custom AutoField throws "useForm must be used within a form"

### ISSUE THREAD

**Issue by ljmotta** [vazco/uniforms#977]
Hi there, I'm using a older version of the library (v3.0), and I've been trying to create my own `AutoField` component. I've followed the [documentation](https://uniforms.tools/docs/examples-custom-fields), and I'm getting the following error: `Error Invariant Violation: useForm must be used within a form.`. Probably the docs reference another version? I've searched the repo and the newer versions has another way to create a custom `AutoField`, but updating is my last option. This is what I've done:

I've created my own bridge, which works fine:
```
export class MyCustomBridge extends JSONSchemaBridge {
  public getType(name: string) {
    const { format: fieldFormat } = super.getField(name);
    if (fieldFormat === "days and time duration") {
      return "days and time duration";
    }
    return super.getType(name);
  }
}
```

and my custom `AutoField` just as the docs:
```
function MyAuto(props: any) {
  const determineComponent = (props: any) => {
    switch (props.fieldType) {
      case "days and time duration":
        return DaysAndTimeDuration;
      default:
        return AutoField;
    }
  };

  const Component = determineComponent(props);

  return <Component {...props} />;
}

export const MyAutoField = connectField(MyAuto,  {
  initialValue: false
});

```

And I've passed it to the `autoField` props of my `AutoForm` component.

```
<AutoForm {...props} autoField={MyAutoField} />
```

I'm missing something? I've tried sevaral setups, but with the same error.

**Comment by wadamek65** [comment]
Hi @ljmotta , thanks for the issue report.

One thing I noticed which does not follow our documentation example is that you do not pass an empty string value as the name property to the component in your custom auto field `<Component {...props} name="" />`. I really doubt this is the cause of your problem though. Other than that I don't see any major issues in the code you provided and after a brief check I couldn't find any related issues that might have affected it after version `3.0`.

But just to make sure, can you please update your `uniforms` version temporarily to the newest one and see if the problem still persists? Or create a playground with minimal reproduction so that we can take a look into it? This would at least give us an idea of where to look at.

**Comment by radekmie** [comment]
Also, @ljmotta, could you paste the entire stack trace of this error? It'd be good to know, which component exactly throws this error. Your code looks fine, but maybe there's some pesky detail.

**Comment by ljmotta** [comment]
Hi @wadamek65 @radekmie  Adding the `name` props didn't change the behavior. The entire stack trace is basically this:
```
Error Invariant Violation: useForm must be used within a form.
    at invariant
    at useForm
    at useField
    at MyAutoField
    at renderWithHooks
    at mountIndeterminateComponent
    at beginWork
    at HTMLUnknownElement.callCallback
    at Object.invokeGuardedCallbackDev
    at invokeGuardedCallback
```

After updating to version `3.5.1`. I've started to use the `createAutoField`. My `MyAutoField` component changed to:
```
export const MyAutoField = createAutoField(props => {
  switch (props.fieldType) {
    case "days and time duration":
      return DaysAndTimeDuration;
    default:
      return AutoField;
  }
});
```
And it works, but only for a `JSON Schema` with one depth of properties.
 - Work:
```
const schema = {
  type: 'object',
  properties: {
    foo: {
      format: "days and time duration",
      type: 'string'
    },
  }
};
```

 - Doens't work:
```
const schema = {
  type: 'object',
  properties: {
    foo: {
      type: 'object',
      properties: {
        bar: {
          format: "days and time duration",
          type: 'string'
        },
      },
    }
  }
};
```

- Stack trace:
```
Uncaught Invariant Violation: Unsupported field type: days and time duration
    at invariant (http://localhost:3000/static/js/0.chunk.js:15242:15)
    at http://localhost:3000/static/js/0.chunk.js:56753:58
    at AutoField (http://localhost:3000/static/js/0.chunk.js:60369:79)
   ...

The above error occurred in the <AutoField> component:
    in AutoField (created by Nest)
    in div (created by Nest)
    in Nest (created by NestField)
    in NestField (created by AutoField)
    in AutoField (created by AutoField)
    in AutoField (created by AutoValidatedQuickBootstrap4Form)
    in form (created by AutoValidatedQuickBootstrap4Form)
    in AutoValidatedQuickBootstrap4Form (at App.js:67)
    in div (at App.js:66)
    in App (at src/index.js:8)
```

Curiously, the same happens with my previous implementation of `MyAutoField` (using the `connectField`) but with another error:
```
Uncaught Invariant Violation: Field not found in schema: "foo.foo"
    at invariant (http://localhost:3000/static/js/0.chunk.js:15242:15)
    at http://localhost:3000/static/js/0.chunk.js:58820:55
    at Array.reduce (<anonymous>)
    at MyCustomBridge.getField (http://localhost:3000/static/js/0.chunk.js:58793:82)
    at MyCustomBridge.memoized [as getField] (http://localhost:3000/static/js/0.chunk.js:22963:23)
    at useField (http://localhost:3000/static/js/0.chunk.js:60623:32)
    at AutoField (http://localhost:3000/static/js/0.chunk.js:60367:89)

The above error occurred in the <AutoField> component:
    in AutoField (at App.js:98)
    in MyAuto (created by MyAutoField)
    in MyAutoField (created by AutoValidatedQuickBootstrap4Form)
    in form (created by AutoValidatedQuickBootstrap4Form)
    in AutoValidatedQuickBootstrap4Form (at App.js:67)
    in div (at App.js:66)
    in App (at src/index.js:8)
```

**Comment by radekmie** [comment]
@ljmotta When it comes to the error you've got after upgrading the package, then it's expected (but not really helpful). If you're using your `MyAutoField` only as the `autoField` prop to your form, it's not going to be passed down to the entire subtree. To make it work, you have to use the `AutoField.componentDetectorContext.Provider`, as described in [docs](https://uniforms.tools/docs/uth-autofield-algorithm#overriding-autofield).

The "`foo.foo` error" should be fixed with `name=""`, as otherwise, the `MyAutoField` passes down its own `name`, effectively doubling it.

And finally, the first error (older package version). Could you create a small snippet of the code you're using that fails? I'm really surprised that only upgrading the package results in such an error.

Also, can you make sure that all of the `uniforms*` packages you are using are in the same version? I see you're using `^` in your `package.json` and it may be a problem. All non-core uniforms packages, e.g., themes and bridges, depending on `uniforms` of at least the same version. Now, let's say there's `uniforms@3.0.0` but `uniforms-bootstrap4@3.0.1`. That means the theme will have its own `uniforms@3.0.1`. That will cause trouble as the instance of `React.context` created by these two packages will be different.

---

Maybe we should deprecate the `autoField` entirely? I'll have to think about it.

**Comment by ljmotta** [comment]
@radekmie Hey, sorry for the delay to reply you. About the `AutoField.componentDectectorContext.Provider`, it works like a charm, thank you. My two cents about the `autoField` property, is to remove it or implement the same functionality of the context. It's kind of strange for the implementation not to be passed down to the entire subtree (for me). Or maybe the [docs](https://uniforms.tools/docs/api-fields/#autofields) should have more details about its functionalities/limitations.

Here is a gist with my code using version `3.0.0`: https://gist.github.com/ljmotta/a608e730621e959f1550df58981225ae
Maybe I'm doing something wrong?

**Comment by radekmie** [comment]
@ljmotta No problem! The `autoField` property is much older and we definitely should reconsider its existence. But in any case, I'm glad you've worked it out.

By the way, we have had the exact same problem in one of our projects, and it turned out to be the versioning - there were two `uniforms` packages in `node_modules`. Getting rid of one solved the problem.

**Comment by radekmie** [comment]
As the problem of this is issue is already resolved and I created #980 to discuss the fate of the `autoField` prop, I'm closing this one.
